File: lab1/CoolFilms/CoolFilms/ViewModel/FavoredViewModel.swift
import Foundation
import FirebaseDatabase
import FirebaseAuth

class FavoredViewModel: ObservableObject {
    private let database = Database.database()
    private let auth = Auth.auth()
    
    @Published var films: [Film] = []
    
    func loadFilms() {
        guard let uid = auth.currentUser?.uid else {
            return
        }
        
        let favoredRef = database.reference().child("favored").child(uid)
        favoredRef.observe(.value) { [weak self] snapshot in
            var favoredFilmIds: [String] = []
            
            for child in snapshot.children {
                if let snapshot = child as? DataSnapshot,
                   let filmId = snapshot.value as? String {
                    favoredFilmIds.append(filmId)
                }
            }
            
            let filmsRef = self?.database.reference().child("films")
            filmsRef?.observe(.value) { [favoredFilmIds] snapshot in
                var films: [Film] = []
                
                for child in snapshot.children {
                    if let snapshot = child as? DataSnapshot,
                       let filmId = snapshot.key as String?,
                       let filmValue = snapshot.value as? [String: Any],
                       favoredFilmIds.contains(filmId) {
                        let film = Film(name: filmValue["name"] as? String ?? "",
                                        categories: filmValue["categories"] as? [String] ?? [],
                                        description: filmValue["description"] as? String ?? "")
                        film.id = filmId
                        films.append(film)
                    }
                }
                
                self?.films = films
            }
        }
    }
}

File: lab3/lib/viewmodel/favored_viewmodel.dart
import 'package:firebase_database/firebase_database.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/cupertino.dart';

import '../model/film.dart';

class FavoredViewModel extends ChangeNotifier {
  final FirebaseDatabase _database = FirebaseDatabase.instance;
  final FirebaseAuth _auth = FirebaseAuth.instance;

  List<Film> _films = [];

  List<Film> get films => _films;

  void loadFilms() {
    final uid = _auth.currentUser?.uid;
    if (uid == null) {
      return;
    }

    final favoredRef = _database.ref('favored').child(uid);
    favoredRef.onValue.listen((event) {
      final favoredFilmIds = <String>[];
      for (final child in event.snapshot.children) {
        final filmId = child.value as String?;
        if (filmId != null) {
          favoredFilmIds.add(filmId);
        }
      }

      final filmsRef = _database.ref('films');
      filmsRef.onValue.listen((event) {
        final films = <Film>[];
        for (final child in event.snapshot.children) {
          final filmId = child.key;
          final filmValue = child.value as Map<dynamic, dynamic>;
          if (filmId != null && favoredFilmIds.contains(filmId)) {
            final film = Film(
                name: filmValue['name'] as String? ?? '',
                categories:
                    List<String>.from(filmValue['categories'] as List? ?? []),
                description: filmValue['description'] as String? ?? '',
                id: filmId);
            films.add(film);
          }
        }
        _films = films;
        notifyListeners();
      });
    });
  }
}
